File: dungeon_gen/room_builder.cpp
#include "architector.h"

room_builder::room_builder (std::vector<std::vector<char>>& map, sf::Vector2i size) {
    this->map  = &map;
    this->size = size;

    dir = 0;
    _alive = false;
}

room_builder::room_builder (std::vector<std::vector<char>>* map, sf::Vector2i size) {
    this->map  = map;
    this->size = size;

    dir = 0;
    _alive = false;
}

std::vector<architector*>* room_builder::step () {
    using namespace std;

    random_device rd{};
    mt19937 gen{rd()};
    normal_distribution<double> gen_normal{10.0, 2.5};

    std::vector<architector*>* res = nullptr;
    sf::Vector2i room_size(max<int>(5, abs(round(gen_normal(gen)))), max<int>(5, abs(round(gen_normal(gen)))));

    int left_l = 0, right_l = 0;
    int left_dir  = (dir + 3) % 4;
    int right_dir = (dir + 1) % 4;

    sf::Vector2i left_pos  = pos + arr_dir[dir] + arr_dir[left_dir];
    sf::Vector2i right_pos = pos + arr_dir[dir];

    // search for free space

    for (int i = 0; i < room_size.x - 1; i++, left_pos += arr_dir[left_dir]) {
        sf::Vector2i buf = left_pos;
        bool clear = true;

        for (int j = 0; j < room_size.y; j++, buf += arr_dir[dir]) {
            if (buf.x < 0       || buf.y < 0 ||
                buf.x >= size.x || buf.y >= size.y) {
                clear = false;
                break;
            }

            if ((*map)[buf.x][buf.y]) {
                clear = false;
                break;
            }
        }

        if (clear) {
            left_l++;
        } else {
            break;
        }
    }

    for (int i = 0; i < room_size.x; i++, right_pos += arr_dir[right_dir]) {
        sf::Vector2i buf = right_pos;
        bool clear = true;

        for (int j = 0; j < room_size.y; j++, buf += arr_dir[dir]) {
            if (buf.x < 0       || buf.y < 0 ||
                buf.x >= size.x || buf.y >= size.y) {
                clear = false;
                break;
            }

            if ((*map)[buf.x][buf.y]) {
                clear = false;
                break;
            }
        }

        if (clear) {
            right_l++;
        } else {
            break;
        }
    }

    if (!right_l || left_l + right_l < room_size.x) {
        kill();
        return res;
    }

    int d = (left_l + right_l - room_size.x) ? (std::abs(rand()) % (left_l + right_l - room_size.x)) : 0;
    left_pos  = pos + arr_dir[dir] + arr_dir[left_dir];
    right_pos = pos + arr_dir[dir];

    if (left_l - d == 0 || room_size.x + d - left_l == 1) {
        kill();
        return res;
    }

    // draw room

    for (int i = 0; i < left_l - d; i++, left_pos += arr_dir[left_dir]) {
        sf::Vector2i buf = left_pos;

        for (int j = 0; j < room_size.y; j++, buf += arr_dir[dir]) {
            if (i == left_l - d - 1 ||
                j == 0 || j == room_size.y - 1) {
                (*map)[buf.x][buf.y] = wall;
            } else {
                (*map)[buf.x][buf.y] = room;
            }
        }
    }

    for (int i = 0; i < room_size.x + d - left_l; i++, right_pos += arr_dir[right_dir]) {
        sf::Vector2i buf = right_pos;

        for (int j = 0; j < room_size.y; j++, buf += arr_dir[dir]) {
            if (i == room_size.x + d - left_l - 1 ||
                j == 0 || j == room_size.y - 1) {
                (*map)[buf.x][buf.y] = wall;
            } else {
                (*map)[buf.x][buf.y] = room;
            }
        }
    }

    // draw door

    sf::Vector2i door_pos = pos + arr_dir[dir];
    (*map)[door_pos.x][door_pos.y] = door;

    // genere new objects

    sf::Vector2i new_doors_pos = door_pos;
    int new_dors_dir = (dir + 1) % 4;
    for (int i = 0; i < 2 * (room_size.x + room_size.y - 4); i++) {
        int bufdir = (new_dors_dir + 3) % 4; // direction to xenter
        sf::Vector2i tocen = new_doors_pos + arr_dir[bufdir];

        if ((*map)[tocen.x][tocen.y] == wall) {
            new_dors_dir = bufdir;
            new_doors_pos += arr_dir[new_dors_dir];
            bufdir = (new_dors_dir + 3) % 4;
        }

        bufdir = (new_dors_dir + 1) % 4;
        if (probability_random(0.1)) {
            set_new_architectors(res, new_doors_pos, bufdir, {4, 10});
            (*map)[new_doors_pos.x][new_doors_pos.y] = door;
        }

        new_doors_pos += arr_dir[new_dors_dir];
    }

    return res;
}



void room_builder::spawn (sf::Vector2i pos, int dir) {
    if (_alive) return;

    _alive = true;

    pos.x = std::abs(pos.x) % size.x;
    pos.y = std::abs(pos.y) % size.y;

    this->dir = std::abs(dir) % 4;
    this->pos = pos;
}

File: underworldProj/Assets/scripts/architectors/RoomBuilder.cs
using System.Collections.Generic;
using tiles;
using UnityEngine;
using UnityEngine.Audio;

public class RoomBuilder : Architector {
    public RoomBuilder (Dungeon dungeon) : base(dungeon) { }

    public override void spawn (Vector3Int pos, Direction dir) {
        this.pos = pos;
        this.dir = dir;
        alive = true;
    }
    
    public override List<Architector> step() {
        var roomSize = new Vector3Int(
            x: Utilites.rand.getInt(3, 5) + Utilites.rand.getInt(3, 5),
            z: Utilites.rand.getInt(3, 5) + Utilites.rand.getInt(3, 5),
            y: 0
        );

        var freeSpace = searchFreeSpace(roomSize);
        int leftL = freeSpace.x, rightL = freeSpace.z;
        var dirLeft  = dir.getLeft();
        var dirRight = dir.getRight();
        
        var leftPos  = pos + dir.getDir() + dirLeft.getDir();
        var rightPos = pos + dir.getDir();
        
        if (rightL == 0 || leftL + rightL < roomSize.x) {
            kill();
            return null;
        }

        var d = (leftL + rightL - roomSize.x > 0) ? (Utilites.rand.getUInt() % (leftL + rightL - roomSize.x)) : 0;
        leftPos  = pos + dir.getDir() + dirLeft.getDir();
        rightPos = pos + dir.getDir();

        if (leftL - d == 0 || roomSize.x + d - leftL == 1) {
            kill();
            return null;
        }
        
        // draw room

        for (var i = 0; i < leftL - d; i++, leftPos += dirLeft.getDir()) {
            var buf = leftPos;

            for (var j = 0; j < roomSize.z; j++, buf += dir.getDir()) {
                if (i == leftL - d - 1 || j == 0 || j == roomSize.z - 1) {
                    dungeon.tileMap[buf.x][buf.z] = new Wall(dungeon);
                } else {
                    dungeon.tileMap[buf.x][buf.z] = new Room(dungeon);
                }
            }
        }
        for (var i = 0; i < roomSize.x + d - leftL; i++, rightPos += dirRight.getDir()) {
            var buf = rightPos;

            for (var j = 0; j < roomSize.z; j++, buf += dir.getDir()) {
                if (i == roomSize.x + d - leftL - 1 || j == 0 || j == roomSize.z - 1) {
                    dungeon.tileMap[buf.x][buf.z] = new Wall(dungeon);
                } else {
                    dungeon.tileMap[buf.x][buf.z] = new Room(dungeon);
                }
            }
        }

        // draw door

        var doorPos = pos + dir.getDir();
        dungeon.tileMap[doorPos.x][doorPos.z] = new Door(dungeon);
        
        // spawn new objects

        var listArch = new List<Architector>();
        
        var aroundDir = dir.getRight();
        var curDoorPos = pos + dir.getDir();
        for (var i = 0; i < 2 * (roomSize.x + roomSize.z - 4); i++) {
            var centerDir = aroundDir.getLeft();
            curDoorPos += aroundDir.getDir();
            
            dungeon.tileMap[curDoorPos.x][curDoorPos.z].setDirection(centerDir);
            if (dungeon.tileMap[(centerDir.getDir() + curDoorPos).x][(centerDir.getDir() + curDoorPos).z].getType()
                == DungeonTile.type.Wall) {
                dungeon.tileMap[curDoorPos.x][curDoorPos.z].setupAngel();
                
                aroundDir.left();
                curDoorPos += aroundDir.getDir();
                continue;
            }

            if (Utilites.rand.getProbabilityBool(0.15f)) {
                listArch.Add(setNewArchitector(curDoorPos, centerDir.getReverse(), new float[] {1, 60}));
                dungeon.tileMap[curDoorPos.x][curDoorPos.z] = new Door(dungeon);
            }
        }

        kill();
        return listArch;
    }

    private Vector3Int searchFreeSpace (Vector3Int roomSize) {
        int leftL = 0, rightL = 0;
        var dirLeft  = dir.getLeft();
        var dirRight = dir.getRight();
        
        var leftPos  = pos + dir.getDir() + dirLeft.getDir();
        var rightPos = pos + dir.getDir();
        
        for (var i = 0; i < roomSize.x - 1; i++, leftPos  += dirLeft.getDir()) {
            var buf = leftPos;
            var isSpaceFree = true;

            for (var j = 0; j < roomSize.z; j++, buf += dir.getDir()) {
                if (buf.x < 0 || buf.z < 0  ||
                    buf.x >= dungeon.size.x || buf.z >= dungeon.size.z) {
                    isSpaceFree = false;
                    break;
                }

                if (dungeon.tileMap[buf.x][buf.z].getType() != DungeonTile.type.Empty &&
                    dungeon.tileMap[buf.x][buf.z].getType() != DungeonTile.type.Ground) {
                    isSpaceFree = false;
                    break;
                }
            }

            if (isSpaceFree) {
                leftL++;
            } else {
                break;
            }
        }
        for (var i = 0; i < roomSize.x;     i++, rightPos += dirRight.getDir()) {
            var buf = rightPos;
            var isSpaceFree = true;

            for (var j = 0; j < roomSize.z; j++, buf += dir.getDir()) {
                if (buf.x < 0 || buf.z < 0  ||
                    buf.x >= dungeon.size.x || 
                    buf.z >= dungeon.size.z) {
                    isSpaceFree = false;
                    break;
                }

                if (dungeon.tileMap[buf.x][buf.z].getType() != DungeonTile.type.Empty &&
                    dungeon.tileMap[buf.x][buf.z].getType() != DungeonTile.type.Ground) {
                    isSpaceFree = false;
                    break;
                }
            }

            if (isSpaceFree) {
                rightL++;
            } else {
                break;
            }
        }
        
        return new Vector3Int(leftL, 0, rightL);
    }
}